### PR TITLE
systemd: Fix systemd-cryptenroll TPM2 + fix systemd-tmpfiles-setup-dev.service

### LINKED ISSUE #167994: systemd-cryptsetup is broken when used with tpm2
### Describe the bug
Plugins built with `systemd` are not available in `cryptsetup`
```
machine # [    5.839874] systemd-cryptsetup[455]: Token 0 unusable for segment 0 with desired keyslot priority 2.
machine # [    5.840623] systemd-cryptsetup[455]: Trying to load /nix/store/l6y9kks5z6ywsxivnld12hh326lbx2yc-cryptsetup-2.4.3/lib/cryptsetup/libcryptsetup-token-systemd-tpm2.so.
machine # [    5.841665] systemd-cryptsetup[455]: /nix/store/l6y9kks5z6ywsxivnld12hh326lbx2yc-cryptsetup-2.4.3/lib/cryptsetup/libcryptsetup-token-systemd-tpm2.so: cannot open shared object file: No such file or directory
machine # [    5.842962] systemd-cryptsetup[455]: No TPM2 metadata enrolled in LUKS2 header or TPM2 support not available, falling back to traditional unlocking.
```

#### Local with TPM2
```bash
systemd-cryptenroll --tpm2-device=list
dd if=/dev/zero of=encrypted.img bs=1 count=0 seek=1G
echo -n lukspass | cryptsetup luksFormat -q encrypted.img -

# confirm it works with cryptsetup and password
sudo cryptsetup luksOpen encrypted.img foo
# enter password: lukspass
sudo cryptsetup luksClose foo

# enroll tpm2
sudo PASSWORD=lukspass systemd-cryptenroll --tpm2-device=auto --tpm2-pcrs=7 encrypted.img
sudo ${pkgs.systemd}/lib/systemd/systemd-cryptsetup attach luks encrypted.img - tpm2-device=auto
```
Password prompt appears:
```bash
Set cipher aes, mode xts-plain64, key size 512 bits for device encrypted.img.
🔐 Please enter passphrase for disk foo: (press TAB for no echo)
```

```bash
#cleanup
sudo systemd-cryptenroll --wipe-slot=tpm2 encrypted.img
```

#### Using tests with `swtpm`
1. enable debug in systemd `boot.kernelParams = [ "systemd.log_level=debug" "systemd.log_target=console" "console=ttyS0,38400" "console=tty1" ];`
2. run the test `nix build .#nixosTests.systemd-cryptenroll` (give it a minute since it spits out lots of debug statement)
3. inspect the log `nix log .#nixosTests.systemd-cryptenroll`

### Expected behavior
Password prompt should not appear

### Additional context
https://github.com/NixOS/nixpkgs/pull/139864 should've been caught this issue

`strace` suggests that the new [systemd feature](https://github.com/systemd/systemd/blob/928c6c00c37cc01860a0a6fec30e8188bb52929b/NEWS#L647-L649) is to blame:
```
1933590 openat(AT_FDCWD, "/nix/store/l6y9kks5z6ywsxivnld12hh326lbx2yc-cryptsetup-2.4.3/lib/cryptsetup/libcryptsetup-token-systemd-tpm2.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
```

It seems like `cryptsetup` expects the [library](https://gitlab.com/cryptsetup/cryptsetup/-/blob/main/lib/luks2/luks2_token.c#L154) to be present under `${cryptsetup}/lib/cryptsetup` but it's present under `${systemd}/lib/cryptsetup`
No matter what I do I get into dependency loop 🤷🏻 

### Notify maintainers

@flokli @kloenk @mic92

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
 - system: `"x86_64-linux"`
 - host os: `Linux 5.17.1, NixOS, 22.05 (Quokka)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.8.0pre20220322_d5d4d98`
 - channels(ymatsiuk): `""`
 - channels(root): `"nixos"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

### PR DESCRIPTION
**Pull request by dasJ**
###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by K900 — APPROVED** [review]
LGTM, this (partially) fixes systemd-in-stage1 on my setup

File: pkgs/os-specific/linux/systemd/default.nix [hunk 1]
```diff
@@ -641,12 +643,6 @@ stdenv.mkDerivation {
   '';
 
   postInstall = ''
-    # sysinit.target: Don't depend on
-    # systemd-tmpfiles-setup.service. This interferes with NixOps's
-    # send-keys feature (since sshd.service depends indirectly on
-    # sysinit.target).
-    mv $out/lib/systemd/system/sysinit.target.wants/systemd-tmpfiles-setup-dev.service $out/lib/systemd/system/multi-user.target.wants/
```
Comment (flokli) [inline pkgs/os-specific/linux/systemd/default.nix:648]: Please put this in an additional PR.

File: pkgs/os-specific/linux/systemd/default.nix [hunk 2]
```diff
@@ -674,7 +670,14 @@ stdenv.mkDerivation {
   preFixup = lib.optionalString withEfi ''
     mv $out/lib/systemd/boot/efi $out/dont-strip-me
   '';
-  postFixup = lib.optionalString withEfi ''
+
+  postFixup = ''
+    # Wrap in the correct path for LUKS2 tokens. Must be after the fixup phase
+    # or the rpath cleanup removes the directories again.
```
Comment (flokli) [inline pkgs/os-specific/linux/systemd/default.nix:676]: This comment doesn't look right. Down there, we set `LD_LIBRARY_PATH`, not set rpath. Also, I'd propose moving the comment above `postFixup`, so iterating on the comment itself doesn't trigger rebuilds.

File: pkgs/os-specific/linux/systemd/default.nix [hunk 3]
```diff
@@ -674,7 +670,14 @@ stdenv.mkDerivation {
...
+    for f in lib/systemd/systemd-cryptsetup bin/systemd-cryptenroll; do
+      wrapProgram $out/$f --prefix LD_LIBRARY_PATH : ${placeholder "out"}/lib/cryptsetup
```
Comment (flokli) [inline pkgs/os-specific/linux/systemd/default.nix:678]: Why do we need to set `LD_LIBRARY_PATH`? I was under the assumption that `${systemd}/lib` being in rpath of all systemd binaries calling out libcryptsetup code is sufficient…

Comment (dasJ) [reply pkgs/os-specific/linux/systemd/default.nix:678]: No, `${systemd}/lib` is not the same thing as `${systemd}/lib/cryptsetup`

Comment (flokli) [reply pkgs/os-specific/linux/systemd/default.nix:678]: So `patchelf --add-rpath`, and disable the rpath trimming?

Comment (ElvishJerricco) [reply pkgs/os-specific/linux/systemd/default.nix:678]: Agreed. `LD_LIBRARY_PATH` is not the solution.

Also, in `systemdMinimal`, these executables don't exist, so the build fails.

**Review by ElvishJerricco — CHANGES_REQUESTED** [review]

**Comment by flokli** [comment]
@dasJ is there a follow-up PR on this?

**Comment by dasJ** [comment]
None planned

File: pkgs/os-specific/linux/systemd/default.nix (hunk 1 above)
Comment (lovesegfault) [reply pkgs/os-specific/linux/systemd/default.nix:648]: Done: https://github.com/NixOS/nixpkgs/pull/171723

**Pull request closed without merging**